src/charm.py

```diff
@@ -6,20 +6,293 @@
+                                                        "key": "username",
+                                                    }
+                                                },
+                                            },
+                                            {
+                                                "name": "DB_PASSWORD",
+                                                "valueFrom": {
+                                                    "secretKeyRef": {
+                                                        "name": f"{self._name}-database-secret",
+                                                        "key": "password",
+                                                    }
+                                                },
+                                            },
+                                            {
+                                                "name": "DB_ENDPOINT",
+                                                "valueFrom": {
+                                                    "secretKeyRef": {
+                                                        "name": f"{self._name}-database-secret",
+                                                        "key": "endpoint",
+                                                    }
+                                                },
+                                            },
+                                            {
+                                                "name": "DSN",
+                                                "value": "postgres://$(DB_USER):$(DB_PASSWORD)@$(DB_ENDPOINT)/postgres",
+                                            },
+                                        ],
+                                    }
+                                ]
+                            }
+                        }
+                    }
+                },
+            )
+        except ApiError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus(
+                f"Applying resources failed with code {str(err.status.code)}."
+            )
+            return
+
+        # Get current layer
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._hydra_layer
+
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration, replanning")
+            try:
+                self._container.replan()
+            except ChangeError as err:
+                logger.error(str(err))
+                self.unit.status = BlockedStatus("Failed to replan")
+                return
+
+        self._run_sql_migration()
+
+    def main(self, event) -> None:
```

Comment:
The single event handler pattern is something that was created in Kubeflow but I do not believe it should be dragged on to other teams. Lets make sure we are using the same patterns in both Kratos and Hyrdra.